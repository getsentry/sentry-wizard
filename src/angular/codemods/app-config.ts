import type { ObjectProperty } from '@babel/types';

// @ts-expect-error - magicast is ESM and TS complains about that. It works though
import type { ProxifiedModule } from 'magicast';

// @ts-expect-error - clack is ESM and TS complains about that. It works though
import * as clack from '@clack/prompts';
import { gte, type SemVer } from 'semver';
import * as recast from 'recast';
import pc from 'picocolors';

export function updateAppConfigMod(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppConfigMod: ProxifiedModule<any>,
  angularVersion: SemVer,
  isTracingEnabled: boolean,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): ProxifiedModule<any> {
  const isAboveAngularV19 = gte(angularVersion, '19.0.0');

  addImports(originalAppConfigMod, isAboveAngularV19, isTracingEnabled);
  addProviders(originalAppConfigMod, isAboveAngularV19, isTracingEnabled);

  return originalAppConfigMod;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function addSentryImport(originalAppConfigMod: ProxifiedModule<any>): void {
  const imports = originalAppConfigMod.imports;
  const hasSentryImport = imports.$items.some(
    (item) => item.from === '@sentry/angular',
  );

  if (!hasSentryImport) {
    imports.$add({
      from: '@sentry/angular',
      imported: '*',
      local: 'Sentry',
    });
  }
}

function addErrorHandlerImport(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppConfigMod: ProxifiedModule<any>,
): void {
  const imports = originalAppConfigMod.imports;
  const hasErrorHandler = imports.$items.some(
    (item) => item.local === 'ErrorHandler' && item.from === '@angular/core',
  );

  if (!hasErrorHandler) {
    imports.$add({
      from: '@angular/core',
      imported: 'ErrorHandler',
      local: 'ErrorHandler',
    });
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function addRouterImport(originalAppConfigMod: ProxifiedModule<any>): void {
  const imports = originalAppConfigMod.imports;
  const hasRouter = imports.$items.some(
    (item) => item.local === 'Router' && item.from === '@angular/router',
  );

  if (!hasRouter) {
    imports.$add({
      from: '@angular/router',
      imported: 'Router',
      local: 'Router',
    });
  }
}

function addMissingImportsV19(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppConfigMod: ProxifiedModule<any>,
): void {
  const imports = originalAppConfigMod.imports;

  const hasProvideAppInitializer = imports.$items.some(
    (item) =>
      item.local === 'provideAppInitializer' && item.from === '@angular/core',
  );

  if (!hasProvideAppInitializer) {
    imports.$add({
      from: '@angular/core',
      imported: 'provideAppInitializer',
      local: 'provideAppInitializer',
    });
  }

  const hasInject = imports.$items.some(
    (item) => item.local === 'inject' && item.from === '@angular/core',
  );

  if (!hasInject) {
    imports.$add({
      from: '@angular/core',
      imported: 'inject',
      local: 'inject',
    });
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function addAppInitializer(originalAppConfigMod: ProxifiedModule<any>): void {
  const imports = originalAppConfigMod.imports;

  const hasAppInitializer = imports.$items.some(
    (item) => item.local === 'APP_INITIALIZER' && item.from === '@angular/core',
  );

  if (!hasAppInitializer) {
    imports.$add({
      from: '@angular/core',
      imported: 'APP_INITIALIZER',
      local: 'APP_INITIALIZER',
    });
  }
}

function addImports(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppConfigMod: ProxifiedModule<any>,
  isAboveAngularV19: boolean,
  isTracingEnabled: boolean,
): void {
  addSentryImport(originalAppConfigMod);
  addErrorHandlerImport(originalAppConfigMod);

  if (isTracingEnabled) {
    addRouterImport(originalAppConfigMod);
  }

  if (isAboveAngularV19) {
    addMissingImportsV19(originalAppConfigMod);
  } else if (isTracingEnabled) {
    addAppInitializer(originalAppConfigMod);
  }
}

function addProviders(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppConfigMod: ProxifiedModule<any>,
  isAboveAngularV19: boolean,
  isTracingEnabled: boolean,
): void {
  const b = recast.types.builders;

  recast.visit(originalAppConfigMod.exports.$ast, {
    visitExportNamedDeclaration(path) {
      if (
        path.node.declaration?.type !== 'VariableDeclaration' ||
        path.node.declaration.declarations[0]?.type !== 'VariableDeclarator' ||
        path.node.declaration.declarations[0].id.type !== 'Identifier' ||
        path.node.declaration.declarations[0].id.name !== 'appConfig' ||
        path.node.declaration.declarations[0].init?.type !==
          'ObjectExpression' ||
        !path.node.declaration.declarations[0].init.properties
      ) {
        return;
      }

      const appConfigProps =
        path.node.declaration.declarations[0].init.properties;

      const providersProperty = appConfigProps.find(
        (prop) =>
          prop?.type === 'ObjectProperty' &&
          prop.key.type === 'Identifier' &&
          prop.key.name === 'providers',
        // type cast is safe because we already check the type in the find condition
      ) as ObjectProperty | undefined;

      const validProviders =
        providersProperty?.value?.type === 'ArrayExpression'
          ? providersProperty.value
          : undefined;

      if (!validProviders) {
        return;
      }

      // Check if there is already an ErrorHandler provider
      const hasErrorHandlerProvider = validProviders?.elements.some(
        (element) =>
          element &&
          element.type === 'ObjectExpression' &&
          element.properties.some(
            (prop) =>
              prop.type === 'ObjectProperty' &&
              prop.key.type === 'Identifier' &&
              prop.key.name === 'provide' &&
              prop.value.type === 'Identifier' &&
              prop.value.name === 'ErrorHandler',
          ),
      );

      // If there is already an ErrorHandler provider, we skip adding it and log a message
      if (hasErrorHandlerProvider) {
        clack.log.warn(`ErrorHandler provider already exists in your app config.
Please refer to the Sentry Angular SDK documentation to combine it manually with Sentry's ErrorHandler.
${pc.underline(
  'https://docs.sentry.io/platforms/javascript/guides/angular/features/error-handler/',
)}
`);
      } else {
        const errorHandlerObject = b.objectExpression([
          b.objectProperty(
            b.identifier('provide'),
            b.identifier('ErrorHandler'),
          ),
          b.objectProperty(
            b.identifier('useValue'),
            b.identifier('Sentry.createErrorHandler()'),
          ),
        ]);

        validProviders.elements.push(
          // @ts-expect-error - errorHandlerObject is an objectExpression
          errorHandlerObject,
        );
      }

      if (isTracingEnabled) {
        const traceServiceObject = b.objectExpression([
          b.objectProperty(
            b.identifier('provide'),
            b.identifier('Sentry.TraceService'),
          ),
          b.objectProperty(
            b.identifier('deps'),
            b.arrayExpression([b.identifier('Router')]),
          ),
        ]);

        // @ts-expect-error - errorHandlerObject is an objectExpression
        validProviders.elements.push(traceServiceObject);

        if (isAboveAngularV19) {
          const provideAppInitializerCall = b.callExpression(
            b.identifier('provideAppInitializer'),
            [
              b.arrowFunctionExpression(
                [],
                b.blockStatement([
                  b.expressionStatement(
                    b.callExpression(b.identifier('inject'), [
                      b.identifier('Sentry.TraceService'),
                    ]),
                  ),
                ]),
              ),
            ],
          );

          // @ts-expect-error - provideAppInitializerCall is an objectExpression
          validProviders.elements.push(provideAppInitializerCall);
        } else {
          const provideAppInitializerObject = b.objectExpression([
            b.objectProperty(
              b.identifier('provide'),
              b.identifier('APP_INITIALIZER'),
            ),
            b.objectProperty(
              b.identifier('useFactory'),
              b.arrowFunctionExpression(
                [],
                b.arrowFunctionExpression([], b.blockStatement([])),
              ),
            ),
            b.objectProperty(
              b.identifier('deps'),
              b.arrayExpression([b.identifier('Sentry.TraceService')]),
            ),
            b.objectProperty(b.identifier('multi'), b.booleanLiteral(true)),
          ]);

          // @ts-expect-error - provideAppInitializerObject is an objectExpression
          validProviders.elements.push(provideAppInitializerObject);
        }
      }

      this.traverse(path);
    },
  });
}
