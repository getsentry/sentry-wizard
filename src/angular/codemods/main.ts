import type { Program } from '@babel/types';

import {
  builders,
  generateCode,
  Proxified,
  type ProxifiedModule,
  // @ts-expect-error - magicast is ESM and TS complains about that. It works though
} from 'magicast';

export function updateAppEntryMod(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppModuleMod: ProxifiedModule<any>,
  dsn: string,
  selectedFeatures: {
    performance: boolean;
    replay: boolean;
    logs: boolean;
  },
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): ProxifiedModule<any> {
  originalAppModuleMod.imports.$add({
    from: '@sentry/angular',
    imported: '*',
    local: 'Sentry',
  });

  insertInitCall(originalAppModuleMod, dsn, selectedFeatures);

  return originalAppModuleMod;
}

export function insertInitCall(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalAppModuleMod: ProxifiedModule<any>,
  dsn: string,
  selectedFeatures: {
    performance: boolean;
    replay: boolean;
    logs: boolean;
  },
): void {
  const initCallArgs = getInitCallArgs(dsn, selectedFeatures);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- builders return Proxified which defaults to any
  const initCall = builders.functionCall('Sentry.init', initCallArgs);
  const originalAppModuleModAst = originalAppModuleMod.$ast as Program;

  const initCallInsertionIndex = getAfterImportsInsertionIndex(
    originalAppModuleModAst,
  );

  originalAppModuleModAst.body.splice(
    initCallInsertionIndex,
    0,
    // @ts-expect-error - string works here because the AST is proxified by magicast
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument -- builders return Proxified which defaults to any.
    generateCode(initCall).code,
  );
}

type InitCallArgs = Record<
  string,
  string | number | boolean | Array<Proxified>
>;

export function getInitCallArgs(
  dsn: string,
  selectedFeatures: {
    performance: boolean;
    replay: boolean;
    logs: boolean;
  },
): InitCallArgs {
  const initCallArgs: InitCallArgs = {
    dsn,
  };

  if (selectedFeatures.replay || selectedFeatures.performance) {
    initCallArgs.integrations = [];

    if (selectedFeatures.performance) {
      initCallArgs.integrations.push(
        builders.functionCall('Sentry.browserTracingIntegration'),
      );
      initCallArgs.tracesSampleRate = 1.0;
    }

    if (selectedFeatures.replay) {
      initCallArgs.integrations.push(
        builders.functionCall('Sentry.replayIntegration'),
      );

      initCallArgs.replaysSessionSampleRate = 0.1;
      initCallArgs.replaysOnErrorSampleRate = 1.0;
    }
  }

  if (selectedFeatures.logs) {
    initCallArgs.enableLogs = true;
  }

  initCallArgs.sendDefaultPii = true;

  return initCallArgs;
}

/**
 * We want to insert the handleError function just after all imports
 */
export function getAfterImportsInsertionIndex(
  originalEntryServerModAST: Program,
): number {
  for (let x = originalEntryServerModAST.body.length - 1; x >= 0; x--) {
    if (originalEntryServerModAST.body[x].type === 'ImportDeclaration') {
      return x + 1;
    }
  }

  return 0;
}
