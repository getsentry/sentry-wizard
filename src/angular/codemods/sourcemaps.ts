// @ts-expect-error - clack is ESM and TS complains about that. It works though
import * as clack from '@clack/prompts';
import * as path from 'path';
import * as fs from 'fs';
import {
  angularJsonTemplate,
  configureAngularSourcemapGenerationFlow,
} from '../../sourcemaps/tools/angular';

interface PartialAngularJson {
  projects?: {
    [key: string]: {
      architect?: {
        build?: {
          configurations?: {
            production?: {
              sourceMap?: boolean;
            };
          };
        };
      };
    };
  };
}

export async function addSourcemapEntryToAngularJSON(): Promise<void> {
  const angularJsonPath = path.join(process.cwd(), 'angular.json');
  const angularJSONFile = fs.readFileSync(angularJsonPath, 'utf-8');
  const angularJson = JSON.parse(angularJSONFile) as PartialAngularJson;

  if (!angularJson || typeof angularJson !== 'object') {
    await configureAngularSourcemapGenerationFlow();
  }

  const projects = Object.keys(angularJson.projects as Record<string, unknown>);

  if (!projects.length) {
    await configureAngularSourcemapGenerationFlow();
  }

  // Emit sourcemaps from all projects in angular.json
  for (const project of projects) {
    const projectConfig = angularJson.projects?.[project] || {};

    if (!projectConfig.architect) {
      projectConfig.architect = {};
    }

    if (!projectConfig.architect.build) {
      projectConfig.architect.build = {};
    }

    if (!projectConfig.architect.build.configurations) {
      projectConfig.architect.build.configurations = {};
    }

    projectConfig.architect.build.configurations.production = {
      sourceMap: true,
    };
  }

  try {
    fs.writeFileSync(angularJsonPath, JSON.stringify(angularJson, null, 2));
  } catch (error) {
    clack.log.error(`Failed to write sourcemap configuration to angular.json`);
    clack.log
      .warn(`Please add the following configuration to your angular.json file:
        ${angularJsonTemplate}`);
  }
}
