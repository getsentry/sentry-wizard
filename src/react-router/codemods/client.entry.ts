/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

import * as recast from 'recast';
import type { namedTypes as t } from 'ast-types';

// @ts-expect-error - clack is ESM and TS complains about that. It works though
import clack from '@clack/prompts';

// @ts-expect-error - magicast is ESM and TS complains about that. It works though
import { loadFile, writeFile } from 'magicast';
import { hasSentryContent } from '../../utils/ast-utils';
import { getSentryInitClientContent } from '../templates';
import { getAfterImportsInsertionIndex } from './utils';

export async function instrumentClientEntry(
  clientEntryPath: string,
  dsn: string,
  enableTracing: boolean,
  enableReplay: boolean,
  enableLogs: boolean,
): Promise<void> {
  const clientEntryAst = await loadFile(clientEntryPath);

  if (hasSentryContent(clientEntryAst.$ast as t.Program)) {
    clack.log.info(`Sentry initialization found in ${clientEntryPath}`);
    return;
  }

  clientEntryAst.imports.$add({
    from: '@sentry/react-router',
    imported: '*',
    local: 'Sentry',
  });

  const initContent = getSentryInitClientContent(
    dsn,
    enableTracing,
    enableReplay,
    enableLogs,
  );

  (clientEntryAst.$ast as t.Program).body.splice(
    getAfterImportsInsertionIndex(clientEntryAst.$ast as t.Program),
    0,
    ...recast.parse(initContent).program.body,
  );

  await writeFile(clientEntryAst.$ast, clientEntryPath);
}
