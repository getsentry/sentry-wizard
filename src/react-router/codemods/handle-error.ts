/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
import * as recast from 'recast';

import {
  ProxifiedModule,
  // @ts-expect-error - magicast is ESM and TS complains about that. It works though
} from 'magicast';

export function instrumentHandleError(entryServerAst: ProxifiedModule): void {
  // Add Sentry handle request and handle error functions
  const handleRequestTemplate = `
const handleRequest = Sentry.createSentryHandleRequest({
  ServerRouter,
  renderToPipeableStream,
  createReadableStreamFromReadable,
});
`;

  const handleErrorTemplate = `
export const handleError = Sentry.createSentryHandleError({
  logErrors: false
});
`;

  // Insert the handle request function
  const handleRequestAst = recast.parse(handleRequestTemplate).program.body[0];
  const handleErrorAst = recast.parse(handleErrorTemplate).program.body[0];

  // Add the imports for React Router server functions
  entryServerAst.imports.$add({
    from: '@react-router/node',
    imported: 'createReadableStreamFromReadable',
    local: 'createReadableStreamFromReadable',
  });

  entryServerAst.imports.$add({
    from: 'react-dom/server',
    imported: 'renderToPipeableStream',
    local: 'renderToPipeableStream',
  });

  entryServerAst.imports.$add({
    from: 'react-router',
    imported: 'ServerRouter',
    local: 'ServerRouter',
  });

  // Find the insertion point after imports
  let insertionIndex = 0;
  recast.visit(entryServerAst.$ast, {
    visitImportDeclaration(path) {
      insertionIndex = Math.max(insertionIndex, path.value.loc?.end?.line || 0);
      this.traverse(path);
    },
  });

  // Insert the handle request and error functions
  recast.visit(entryServerAst.$ast, {
    visitProgram(path) {
      path.value.body.push(handleRequestAst);
      path.value.body.push(handleErrorAst);
      this.traverse(path);
    },
  });

  // Replace default export with handleRequest
  recast.visit(entryServerAst.$ast, {
    visitExportDefaultDeclaration(path) {
      path.value.declaration =
        recast.types.builders.identifier('handleRequest');
      this.traverse(path);
    },
  });
}
