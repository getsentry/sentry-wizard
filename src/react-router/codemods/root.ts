/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
import * as recast from 'recast';
import * as path from 'path';

import type { ExportNamedDeclaration } from '@babel/types';
import type { namedTypes as t } from 'ast-types';

import {
  loadFile,
  writeFile,
  // @ts-expect-error - magicast is ESM and TS complains about that. It works though
} from 'magicast';

import { ERROR_BOUNDARY_TEMPLATE } from '../templates';
import { hasSentryContent } from '../../utils/ast-utils';
import { debug } from '../../utils/debug';

export async function instrumentRoot(rootFileName: string): Promise<void> {
  const rootRouteAst = await loadFile(
    path.join(process.cwd(), 'app', rootFileName),
  );

  const exportsAst = rootRouteAst.exports.$ast as t.Program;

  const namedExports = exportsAst.body.filter(
    (node) => node.type === 'ExportNamedDeclaration',
  ) as ExportNamedDeclaration[];

  let foundErrorBoundary = false;

  namedExports.forEach((namedExport) => {
    const declaration = namedExport.declaration;

    if (!declaration) {
      return;
    }

    if (declaration.type === 'FunctionDeclaration') {
      if (declaration.id?.name === 'ErrorBoundary') {
        foundErrorBoundary = true;
      }
    } else if (declaration.type === 'VariableDeclaration') {
      const declarations = declaration.declarations;

      declarations.forEach((declaration) => {
        // @ts-expect-error - id should always have a name in this case
        if (declaration.id?.name === 'ErrorBoundary') {
          foundErrorBoundary = true;
        }
      });
    }
  });

  const alreadyHasSentry = hasSentryContent(rootRouteAst.$ast as t.Program);

  if (!alreadyHasSentry) {
    rootRouteAst.imports.$add({
      from: '@sentry/react-router',
      imported: '*',
      local: 'Sentry',
    });
  }

  if (!foundErrorBoundary) {
    // Check if `isRouteErrorResponse` is imported, as it's needed in our ErrorBoundary template
    const hasIsRouteErrorResponseImport = rootRouteAst.imports.$items.some(
      (item) =>
        item.imported === 'isRouteErrorResponse' &&
        item.from === 'react-router',
    );

    if (!hasIsRouteErrorResponseImport) {
      rootRouteAst.imports.$add({
        from: 'react-router',
        imported: 'isRouteErrorResponse',
        local: 'isRouteErrorResponse',
      });
    }

    recast.visit(rootRouteAst.$ast, {
      visitExportDefaultDeclaration(path) {
        const implementation = recast.parse(ERROR_BOUNDARY_TEMPLATE).program
          .body[0];

        path.insertBefore(
          recast.types.builders.exportDeclaration(false, implementation),
        );

        this.traverse(path);
      },
    });
  } else {
    let hasBlockStatementBody = false;
    let hasFunctionDeclarationBody = false;

    recast.visit(rootRouteAst.$ast, {
      visitExportNamedDeclaration(path) {
        // Find ErrorBoundary export
        if (
          path.value.declaration?.declarations?.[0].id?.name === 'ErrorBoundary'
        ) {
          hasBlockStatementBody = true;
        }

        if (path.value.declaration?.id?.name === 'ErrorBoundary') {
          hasFunctionDeclarationBody = true;
        }

        if (hasBlockStatementBody || hasFunctionDeclarationBody) {
          const errorBoundaryExport = hasBlockStatementBody
            ? path.value.declaration?.declarations?.[0].init
            : path.value.declaration;

          let alreadyHasCaptureException = false;

          // Check if `Sentry.captureException` or `captureException` is already called inside the ErrorBoundary
          recast.visit(errorBoundaryExport, {
            visitCallExpression(callPath) {
              const callee = callPath.value.callee;
              if (
                (callee.type === 'MemberExpression' &&
                  callee.object.name === 'Sentry' &&
                  callee.property.name === 'captureException') ||
                (callee.type === 'Identifier' &&
                  callee.name === 'captureException')
              ) {
                alreadyHasCaptureException = true;
              }

              this.traverse(callPath);
            },
          });

          if (!alreadyHasCaptureException) {
            // Add Sentry.captureException call
            const captureExceptionCall = recast.parse(
              `Sentry.captureException(error);`,
            ).program.body[0];

            // Check whether ErrorBoundary is a function declaration or variable declaration
            const isFunctionDeclaration =
              errorBoundaryExport.type === 'FunctionDeclaration';
            const isVariableDeclaration =
              errorBoundaryExport.type === 'VariableDeclaration';

            if (isFunctionDeclaration) {
              // If it's a function declaration, we can insert the call directly
              errorBoundaryExport.body.body.splice(
                errorBoundaryExport.body.body.length - 1,
                0,
                captureExceptionCall,
              );
            } else if (isVariableDeclaration) {
              // If it's a variable declaration, we need to find the right place to insert the call
              const init = errorBoundaryExport.init;
              if (
                init &&
                (init.type === 'ArrowFunctionExpression' ||
                  init.type === 'FunctionExpression')
              ) {
                init.body.body.splice(
                  init.body.body.length - 1,
                  0,
                  captureExceptionCall,
                );
              }
            }
          }
        }

        this.traverse(path);
      },
    });
  }

  await writeFile(
    rootRouteAst.$ast,
    path.join(process.cwd(), 'app', rootFileName),
  );
}
