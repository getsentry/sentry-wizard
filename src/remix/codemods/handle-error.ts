/* eslint-disable @typescript-eslint/no-unsafe-assignment */

// @ts-expect-error - magicast is ESM and TS complains about that. It works though
import type { ProxifiedModule } from 'magicast';
import type { Program } from '@babel/types';

import * as recast from 'recast';

import { HANDLE_ERROR_TEMPLATE_V2 } from '../templates';
import { getInitCallInsertionIndex, hasSentryContent } from '../utils';

// @ts-expect-error - clack is ESM and TS complains about that. It works though
import clack from '@clack/prompts';
import chalk from 'chalk';

// @ts-expect-error - magicast is ESM and TS complains about that. It works though
import { generateCode } from 'magicast';

export function instrumentHandleError(
  originalEntryServerMod: ProxifiedModule<any>,
  serverEntryFilename: string,
): boolean {
  const originalEntryServerModAST = originalEntryServerMod.$ast as Program;

  const handleErrorFunction = originalEntryServerModAST.body.find(
    (node) =>
      node.type === 'ExportNamedDeclaration' &&
      node.declaration?.type === 'FunctionDeclaration' &&
      node.declaration.id?.name === 'handleError',
  );

  if (!handleErrorFunction) {
    clack.log.warn(
      `Could not find function ${chalk.cyan('handleError')} in ${chalk.cyan(
        serverEntryFilename,
      )}. Creating one for you.`,
    );

    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const implementation = recast.parse(HANDLE_ERROR_TEMPLATE_V2).program
      .body[0];

    originalEntryServerModAST.body.splice(
      getInitCallInsertionIndex(originalEntryServerModAST),
      0,
      // @ts-expect-error - string works here because the AST is proxified by magicast
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      recast.types.builders.exportNamedDeclaration(implementation),
    );
  } else if (
    hasSentryContent(
      generateCode(handleErrorFunction).code,
      originalEntryServerMod.$code,
    )
  ) {
    return false;
  } else {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const implementation = recast.parse(HANDLE_ERROR_TEMPLATE_V2).program
      .body[0];

    // @ts-expect-error - string works here because the AST is proxified by magicast
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
    handleErrorFunction.declaration.body.body.unshift(
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      recast.parse(HANDLE_ERROR_TEMPLATE_V2).program.body[0].body.body[0],
    );

    // First parameter is the error
    //
    // @ts-expect-error - string works here because the AST is proxified by magicast
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    handleErrorFunction.declaration.params[0] = implementation.params[0];

    // Second parameter is the request inside an object
    // Merging the object properties to make sure it includes request
    //
    // @ts-expect-error - string works here because the AST is proxified by magicast
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    if (handleErrorFunction.declaration.params?.[1]?.properties) {
      // @ts-expect-error - string works here because the AST is proxified by magicast
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      handleErrorFunction.declaration.params[1].properties.push(
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        implementation.params[1].properties[0],
      );
    } else {
      // Create second parameter if it doesn't exist
      //
      // @ts-expect-error - string works here because the AST is proxified by magicast
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      handleErrorFunction.declaration.params[1] = implementation.params[1];
    }
  }

  return true;
}
