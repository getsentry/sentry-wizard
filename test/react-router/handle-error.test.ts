// @ts-expect-error - magicast is ESM and TS complains about that. It works though
import { parseModule } from 'magicast';
import { describe, expect, it } from 'vitest';
import { instrumentHandleError } from '../../src/react-router/codemods/handle-error';

describe('React Router Handle Error Codemod', () => {
  describe('instrumentHandleError', () => {
    it('should add Sentry handle request and error functions to empty server entry', () => {
      const entryServerAst = parseModule('');

      instrumentHandleError(entryServerAst);

      const result = entryServerAst.generate().code;

      expect(result).toContain(
        'import {  createReadableStreamFromReadable,} from "@react-router/node"',
      );
      expect(result).toContain(
        'import {  renderToPipeableStream,} from "react-dom/server"',
      );
      expect(result).toContain('import {  ServerRouter,} from "react-router"');
      expect(result).toContain(
        'const handleRequest = Sentry.createSentryHandleRequest',
      );
      expect(result).toContain(
        'export const handleError = Sentry.createSentryHandleError',
      );
    });

    it('should add Sentry functions to server entry with existing imports', () => {
      const entryServerAst = parseModule(`
        import { createRequestHandler } from '@react-router/node';
        import express from 'express';

        const app = express();
      `);

      instrumentHandleError(entryServerAst);

      const result = entryServerAst.generate().code;

      expect(result).toContain(
        'import {createRequestHandler, createReadableStreamFromReadable} from \'@react-router/node\'',
      );
      expect(result).toContain('import express from \'express\'');
      expect(result).toContain(
        'import {renderToPipeableStream} from \'react-dom/server\'',
      );
      expect(result).toContain('import {ServerRouter} from \'react-router\'');
      expect(result).toContain(
        'const handleRequest = Sentry.createSentryHandleRequest',
      );
      expect(result).toContain(
        'export const handleError = Sentry.createSentryHandleError',
      );
    });

    it('should replace existing default export with handleRequest', () => {
      const entryServerAst = parseModule(`
        import { createRequestHandler } from '@react-router/node';

        const handler = createRequestHandler({
          build: require('./build'),
        });

        export default handler;
      `);

      instrumentHandleError(entryServerAst);

      const result = entryServerAst.generate().code;

      expect(result).toContain('export default handleRequest');
      expect(result).not.toContain('export default handler');
    });

    it('should handle server entry with function default export', () => {
      const entryServerAst = parseModule(`
        import { createRequestHandler } from '@react-router/node';

        export default function handler(request, response) {
          return createRequestHandler({
            build: require('./build'),
          })(request, response);
        }
      `);

      instrumentHandleError(entryServerAst);

      const result = entryServerAst.generate().code;

      expect(result).toContain(
        'const handleRequest = Sentry.createSentryHandleRequest',
      );
      expect(result).toContain(
        'export const handleError = Sentry.createSentryHandleError',
      );
      expect(result).toContain('export default handleRequest');
    });

    it('should add proper Sentry handle request configuration', () => {
      const entryServerAst = parseModule('');

      instrumentHandleError(entryServerAst);

      const result = entryServerAst.generate().code;

      expect(result).toMatchSnapshot();
    });

    it('should add proper Sentry handle error configuration', () => {
      const entryServerAst = parseModule('');

      instrumentHandleError(entryServerAst);

      const result = entryServerAst.generate().code;

      expect(result).toContain(
        'export const handleError = Sentry.createSentryHandleError({',
      );
      expect(result).toContain('logErrors: false');
    });
  });
});
