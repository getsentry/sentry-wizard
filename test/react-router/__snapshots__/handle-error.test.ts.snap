// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`React Router Handle Error Codemod > instrumentHandleError > should add proper Sentry handle request configuration 1`] = `
"import {  ServerRouter,} from "react-router";
import {  renderToPipeableStream,} from "react-dom/server";
import {  createReadableStreamFromReadable,} from "@react-router/node";
const handleRequest = Sentry.createSentryHandleRequest({
  ServerRouter,
  renderToPipeableStream,
  createReadableStreamFromReadable,
});
export const handleError = Sentry.createSentryHandleError({
  logErrors: false
});"
`;
